[R-1] Update E2E tests for setup flow with Ads setup in Step 1

joemcgill opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

This updates the E2E tests for the full set of requirements for #2215, including the new acceptance flow implemented in #2270.

This adds tests to cover the new modal flow for accepting a new account after creation, and adjusts the tests for steps 3 and 4 to more accurately reflect the new behavior.

### Changelog entry

joemcgill committed:
Add tests for Ads account claim modal

joemcgill committed:
Merge branch 'update/2262-ads-accept-invite-flow' into update/2262-invite-modal-tests

joemcgill committed:
Fix add-paid-campaigns tests

joemcgill committed:
Fix race condition

joemcgill committed:
Fix billing setup tests

joemcgill committed:
Update e2e tests with API changes for status

joemcgill committed:
E2E: Update step 3 & 4 text

joemcgill committed:
E2E: Fix Step 4 tests

joemcgill commented on tests/e2e/utils/pages/setup-mc/step-4-complete-campaign.js:
```diff
@@ -88,16 +88,7 @@ export default class CompleteCampaign extends MockRequests {
 	 * @return {import('@playwright/test').Locator} Get budget section.
 	 */
 	getBudgetSection() {
-		return this.getSections().nth( 4 );
```
I do not understand why this method stopped working because the markup structure seems unchanged. Searching for the specific className for the budget section seems non-ideal, but does now find the correct component.

eason9487 replied:
> I do not understand why this method stopped working because the markup structure seems unchanged.

Its nth index is changed to 2 after removing two `Section` components.

> Searching for the specific className for the budget section seems non-ideal, but does now find the correct component.

Looks good, thanks!

joemcgill commented:
@eason9487 this fixes up all of the E2E tests related to #2205 and #2270. See: https://github.com/woocommerce/google-listings-and-ads/pull/2270#issuecomment-2008467286

I'm happy to merge them directly into the branch for #2270 if you'd prefer to have them all in one place.

eason9487 commented on tests/e2e/specs/add-paid-campaigns/add-paid-campaigns.test.js:
```diff
@@ -175,13 +175,18 @@ test.describe( 'Set up Ads account', () => {
+			await setupAdsAccounts.fulfillAdsAccountStatus( {
+				has_access: true,
+				invite_link: '',
+			} );
```
The `has_access` won't be `true` after an account is just newly created.

eason9487 commented on tests/e2e/specs/add-paid-campaigns/add-paid-campaigns.test.js:
```diff
@@ -206,6 +211,12 @@ test.describe( 'Set up Ads account', () => {
+			await setupAdsAccounts.fulfillAdsAccountStatus( {
+				has_access: true,
+				invite_link: '',
+			} );
```
The above route mocks Google Ads account as disconnected but this route mocks it still has access. This makes the mocked states less close to the actual situation.

eason9487 commented on tests/e2e/utils/mock-requests.js:
```diff
@@ -455,6 +468,43 @@ export default class MockRequests {
+		await this.fulfillAdsAccountStatus( {
+			has_access: false,
+			invite_link: '',
+		} );
```
The `ads/account-status` API won't respond to any content when there is no connected account.

joemcgill replied:
This is no longer true since bb66dd4.

eason9487 commented on tests/e2e/utils/pages/setup-ads/setup-ads-accounts.js:
```diff
@@ -54,6 +54,15 @@ export default class SetupAdsAccount extends MockRequests {
+	/**
+	 * Get the accept account modal.
+	 *
+	 * @return {import('@playwright/test').Locator} The create account modal.
```
The description seems incorrect.

eason9487 commented on tests/e2e/specs/setup-mc/step-1-accounts.test.js:
```diff
@@ -238,20 +250,56 @@ test.describe( 'Set up accounts', () => {
+
+			test.describe( 'Create a claimed account', () => {
+				test.beforeAll( async () => {
+					// Start with no Ads account connected.
```
This code comment seems incorrect.

eason9487 commented on tests/e2e/utils/pages/setup-mc/step-1-set-up-accounts.js:
```diff
@@ -409,13 +409,32 @@ export default class SetUpAccountsPage extends MockRequests {
+			name: 'Create account',
+			exact: true,
+		} );
+		await button.isVisible();
```
This line probably is not required.

eason9487 replied:
Is there a reason this line is required?

joemcgill replied:
Doesn't seem to be. Probably being overly cautious. I've removed in [f7d71b4](https://github.com/woocommerce/google-listings-and-ads/pull/2308/commits/f7d71b42ddaede255326f42e230a764163c59465).

eason9487 commented on tests/e2e/utils/pages/setup-mc/step-1-set-up-accounts.js:
```diff
@@ -409,13 +409,32 @@ export default class SetUpAccountsPage extends MockRequests {
+	 * Click claim account button.
+	 *
+	 * @return {Promise<void>}
+	 */
```
The description and `@return` type seems incorrect.

eason9487 commented on tests/e2e/utils/pages/setup-mc/step-1-set-up-accounts.js:
```diff
@@ -409,13 +409,32 @@ export default class SetUpAccountsPage extends MockRequests {
+	 *
+	 * @return {Promise<void>}
+	 */
+	async getClaimAdsAccountButton() {
```
This function should not need to be async.

eason9487 commented on tests/e2e/utils/pages/setup-mc/step-1-set-up-accounts.js:
```diff
@@ -428,4 +447,35 @@ export default class SetUpAccountsPage extends MockRequests {
+	 * @return {import('@playwright/test').Locator} Get ads account Claim Account button.
+	 */
+	getAdsClaimAccountButton() {
+		return this.getClaimAdsAccountButton();
```
This method is duplicated in the following.
https://github.com/woocommerce/google-listings-and-ads/blob/bb3a22209f363ade918413e39b2344e42b83e0ec/tests/e2e/utils/pages/setup-mc/step-1-set-up-accounts.js#L417-L424

What about leaving this one and removing another one?

```suggestion
		return this.page.getByRole( 'button', {
			name: 'Claim Account',
			exact: true,
		} );
```

eason9487 reviewed (commented):
Thanks for the updates to E2E test. I left some comments and a few of them are blockers since they seems to be incorrect states.

Furthermore, because the E2E test failed while viewing #2205, I haven't looked into the changes to E2E test code. As per https://github.com/woocommerce/google-listings-and-ads/pull/2205#issuecomment-1997559058, I'm going to leave my findings here.

1. Could be removed:
   - https://github.com/woocommerce/google-listings-and-ads/blob/bb3a22209f363ade918413e39b2344e42b83e0ec/tests/e2e/specs/setup-mc/step-1-accounts.test.js#L257-L262
   - https://github.com/woocommerce/google-listings-and-ads/blob/bb3a22209f363ade918413e39b2344e42b83e0ec/tests/e2e/utils/pages/setup-mc/step-1-set-up-accounts.js#L398-L410
   - https://github.com/woocommerce/google-listings-and-ads/blob/bb3a22209f363ade918413e39b2344e42b83e0ec/tests/e2e/utils/pages/setup-mc/step-4-complete-campaign.js#L58-L65
   - https://github.com/woocommerce/google-listings-and-ads/blob/bb3a22209f363ade918413e39b2344e42b83e0ec/tests/e2e/utils/pages/setup-mc/step-4-complete-campaign.js#L67-L74
   - https://github.com/woocommerce/google-listings-and-ads/blob/bb3a22209f363ade918413e39b2344e42b83e0ec/tests/e2e/utils/pages/setup-mc/step-4-complete-campaign.js#L142-L152
   - https://github.com/woocommerce/google-listings-and-ads/blob/bb3a22209f363ade918413e39b2344e42b83e0ec/tests/e2e/utils/pages/setup-mc/step-4-complete-campaign.js#L154-L161
   - https://github.com/woocommerce/google-listings-and-ads/blob/bb3a22209f363ade918413e39b2344e42b83e0ec/tests/e2e/utils/pages/setup-mc/step-4-complete-campaign.js#L196-L205
2. Issues to be addressed:
   - https://github.com/woocommerce/google-listings-and-ads/blob/bb3a22209f363ade918413e39b2344e42b83e0ec/tests/e2e/specs/setup-mc/step-4-complete-campaign.test.js#L200-L204
      - The code comment is incorrect.
      - The regex `/wcdl-section/` is not ideal because both `/wcdl-section/.test('wcdl-section')` and `/wcdl-section/.test('wcdl-section--is-disabled')` are true.
   - https://github.com/woocommerce/google-listings-and-ads/blob/bb3a22209f363ade918413e39b2344e42b83e0ec/tests/e2e/utils/pages/setup-mc/step-4-complete-campaign.js#L81-L83
      - The index in the `.nth` call should be `1`.
3. Descriptions:
   - https://github.com/woocommerce/google-listings-and-ads/blob/bb3a22209f363ade918413e39b2344e42b83e0ec/tests/e2e/specs/setup-mc/step-4-complete-campaign.test.js#L212
      - The "both" in the description doesn't seem accurate.
   - https://github.com/woocommerce/google-listings-and-ads/blob/bb3a22209f363ade918413e39b2344e42b83e0ec/tests/e2e/specs/setup-mc/step-1-accounts.test.js#L762
      - The "without Ads" in the description seem incorrect.
   - https://github.com/woocommerce/google-listings-and-ads/blob/bb3a22209f363ade918413e39b2344e42b83e0ec/tests/e2e/specs/setup-mc/step-1-accounts.test.js#L777
      - The "without MC" in the description seem incorrect.

joemcgill committed:
Merge branch 'update/2262-ads-accept-invite-flow' into update/2262-invite-modal-tests

ankitrox committed:
Fix issues in e2e and remove duplication

ankitrox committed:
Js lint fix

joemcgill committed:
Merge branch 'update/2262-ads-accept-invite-flow' into update/2262-invite-modal-tests

joemcgill committed:
E2E: Fix account status shape in mocks

joemcgill committed:
E2E update setup-mc step 4 tests

joemcgill committed:
Fix continue button during ads setup

joemcgill committed:
E2E improve mocking of account status tests

joemcgill committed:
E2E docs improvements

joemcgill committed:
E2E: Fix test descriptions

joemcgill commented:
@eason9487 I've merged in the upstream changes from #2270 and updated the E2E tests to resolve your earlier feedback. I think this is ready for another look. Just a heads up, I've tried running these again using the GH Action runner, but I think there is something related to the  GTag events tests that has been fixed in develop that have not yet been merged back to this branch, but all tests related to the updates to the changes from #2270 are passing for me locally.

eason9487 commented on tests/e2e/specs/setup-mc/step-1-accounts.test.js:
```diff
@@ -238,23 +250,56 @@ test.describe( 'Set up accounts', () => {
+				() => {
+					test.beforeAll( async () => {
+						// Start with no Ads account connected.
+						await setupAdsAccountPage.mockAdsStatusClaimed();
```
Refer to https://github.com/woocommerce/google-listings-and-ads/pull/2308#discussion_r1537246181. This code comment still seems incorrect.

eason9487 commented on tests/e2e/specs/setup-mc/step-4-complete-campaign.test.js:
```diff
@@ -189,33 +182,24 @@ test.describe( 'Complete your campaign', () => {
+					// check if the section has the class.
+					await expect(
+						adsAudienceSection.locator( 'h1' )
+					).toContainText( 'Ads audience' );
```
The code comment doesn't match the actual assertion.

eason9487 commented on tests/e2e/utils/pages/setup-ads/setup-ads-accounts.js:
```diff
@@ -77,6 +86,55 @@ export default class SetupAdsAccount extends MockRequests {
+	/**
+	 * Get Claim button in the Accept Invitation modal.
+	 *
+	 * @return {import('@playwright/test').Locator} Claim account in modal button.
```
The description seems very confusing.

eason9487 commented on tests/e2e/utils/pages/setup-ads/setup-ads-accounts.js:
```diff
@@ -77,6 +86,55 @@ export default class SetupAdsAccount extends MockRequests {
+	}
+
+	/**
+	 * Click claim account button.
```
The description is incorrect.

eason9487 commented on tests/e2e/utils/pages/setup-ads/setup-ads-accounts.js:
```diff
@@ -77,6 +86,55 @@ export default class SetupAdsAccount extends MockRequests {
+	/**
+	 * Click claim account button.
+	 *
+	 * @return {import('@playwright/test').Locator} Claim account in card button.
```
The description seems very confusing.

eason9487 commented on tests/e2e/utils/pages/setup-ads/setup-ads-accounts.js:
```diff
@@ -77,6 +86,55 @@ export default class SetupAdsAccount extends MockRequests {
+				name: 'Claim Account',
+				exact: true,
+			} )
+			.first();
```
The `.first()` is not required.

eason9487 commented on tests/e2e/utils/pages/setup-ads/setup-ads-accounts.js:
```diff
@@ -77,6 +86,55 @@ export default class SetupAdsAccount extends MockRequests {
+	/**
+	 * Get ads claim account text.
+	 *
+	 * @return {import('@playwright/test').Locator} Get ads account Claim account text.
```
The description seems very confusing.

eason9487 commented on tests/e2e/specs/add-paid-campaigns/add-paid-campaigns.test.js:
```diff
@@ -165,34 +165,68 @@ test.describe( 'Set up Ads account', () => {
 
-		test( 'Create Ads account', async () => {
-			//Intercept Ads connection request
+		test( 'Create and Ads account', async () => {
```
Is there a typo?

eason9487 commented on tests/e2e/specs/add-paid-campaigns/add-paid-campaigns.test.js:
```diff
@@ -165,34 +165,68 @@ test.describe( 'Set up Ads account', () => {
 
 			await connectAdsAccountRequest;
 
+			const modal = await setupAdsAccounts.getAcceptAccountModal();
```
The `await` is not required.

eason9487 commented on tests/e2e/specs/add-paid-campaigns/add-paid-campaigns.test.js:
```diff
@@ -165,34 +165,68 @@ test.describe( 'Set up Ads account', () => {
+		test( 'Show Unclaimed Ads account', async () => {
+			await setupAdsAccounts.clickCloseAcceptAccountButtonFromModal();
+
+			const claimButton =
+				await setupAdsAccounts.getAdsClaimAccountButton();
+			const claimText = await setupAdsAccounts.getAdsClaimAccountText();
```
These `await` are not required.

eason9487 commented on tests/e2e/utils/mock-requests.js:
```diff
@@ -442,6 +455,45 @@ export default class MockRequests {
 	}
 
+	/**
+	 * Mock Google Ads status when disconnected.
```
Perhaps it would be clearer if replace "when" with "as"?

eason9487 commented on tests/e2e/utils/pages/setup-mc/step-1-set-up-accounts.js:
```diff
@@ -428,4 +419,15 @@ export default class SetUpAccountsPage extends MockRequests {
+	/**
+	 * Get Ads account connected notice text.
+	 *
+	 * @return {import('@playwright/test').Locator} Get ads account connected text.
```
It could be more consistent a little bit.

```suggestion
	 * @return {import('@playwright/test').Locator} Get ads account connected notice text.
```

eason9487 reviewed (commented):
Thanks for the update.

Most of the concerns have been addressed but a few may still be missing. In addition, some new changes add incorrect or confusing code comments, and a few `await` and `.first()` are not unnecessary.

joemcgill committed:
Address review feedback

joemcgill commented:
Thanks @eason9487. Addressed all of your feedback in [f7d71b4](https://github.com/woocommerce/google-listings-and-ads/pull/2308/commits/f7d71b42ddaede255326f42e230a764163c59465). Hopefully the description updates are a bit less confusing.

joemcgill committed:
Merge branch 'update/2262-ads-accept-invite-flow' into update/2262-invite-modal-tests

eason9487 reviewed (commented):
Thanks for the updates.

There's still one last unresolved issue: https://github.com/woocommerce/google-listings-and-ads/pull/2308#discussion_r1557096916.

joemcgill committed:
Update inline code comment

joemcgill commented:
@eason9487 

> There's still one last unresolved issue: https://github.com/woocommerce/google-listings-and-ads/pull/2308#discussion_r1557096916.

Addressed in [d63edd4](https://github.com/woocommerce/google-listings-and-ads/pull/2308/commits/d63edd44a133368b428655b54ebc3b585baef391).

eason9487 reviewed (approved):
Thanks for the work. This PR is ready for merging.
